pkg/logs/input/docker/tailer.go

```diff
@@ -127,15 +128,30 @@ func (t *Tailer) tail(since string) error {
 // readForever reads from the reader as fast as it can,
 // and sleeps when there is nothing to read
 func (t *Tailer) readForever() {
+	c := make(chan struct{}, 1)
 	defer t.decoder.Stop()
 	for {
+		var n int
+		var err error
+
 		select {
 		case <-t.stop:
 			// stop reading new logs from container
 			return
 		default:
 			inBuf := make([]byte, 4096)
-			n, err := t.reader.Read(inBuf)
+			// Read can be blocking because it's a wrapper over an HTTP call
+			// If the Read() timeouts, the tailer will be restarted.
+			go func() {
+				n, err = t.reader.Read(inBuf)
```

Comment:
you should isolate the new logic to read in a dedicated method that should return either the content or an error:
```
func read(reader) ([]byte, error)
```
or that would enable just to change line `138`
```
func read(reader, buffer) (n, error)
```